### PR TITLE
feat: json-formatted output for status commands

### PR DESCRIPTION
**Pull request by strophy**
## Issue being fixed or feature implemented

We plan to implement tests on dashmate, which requires machine-readable output. This PR enables the `--format=json` flag for all `status` commands.

## What was done?

- Implement `json` format flag
- `strip-ansi` colors from `chalk`ed table output
- Refactor status commands

## How Has This Been Tested?

Locally against a testnet node, still waiting for sync

## Breaking Changes

None

## Checklist:

- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have added or updated relevant unit/integration/functional/e2e tests
- [x] I have made corresponding changes to the documentation

**For repository code-owners and collaborators only**
- [x] I have assigned this pull request to a milestone

### CODE REVIEW

File: src/commands/status/core.js [hunk 1]
```diff
@@ -172,28 +173,44 @@ class CoreStatusCommand extends ConfigBaseCommand {
...
+    let output;
+
+    if (getFormat(flags) === OUTPUT_FORMATS.JSON) {
+      Object.keys(outputRows).forEach((key) => {
+        outputRows[key] = stripAnsi(outputRows[key]);
+      });
+      output = JSON.stringify(outputRows);
+    } else {
+      const rows = [];
+      Object.keys(outputRows).forEach((key) => {
```
Comment (shumkov) [inline src/commands/status/core.js]: I would recommend to use `Object.entries`

File: src/constants.js [hunk 2]
```diff
@@ -30,6 +30,10 @@ const HOME_DIR_PATH = process.env.DASHMATE_HOME_DIR
   : path.resolve(os.homedir(), '.dashmate');
 const CONFIG_FILE_PATH = path.join(HOME_DIR_PATH, 'config.json');
 
+const OUTPUT_FORMATS = {
+  JSON: 'json',
```
Comment (shumkov) [inline src/constants.js:34]: Let's add plain format here as well

File: src/oclif/command/BaseCommand.js [hunk 3]
```diff
@@ -129,6 +129,7 @@ BaseCommand.flags = {
     description: 'use verbose mode for output',
     default: false,
   }),
+  format: flagTypes.string(),
```
Comment (shumkov) [inline src/oclif/command/BaseCommand.js]: I think it must be enum: `plain` (by default) and `json`. Please add a description too.

File: src/status/outputStatusOverviewFactory.js [hunk 4]
```diff
@@ -24,9 +27,7 @@ function outputStatusOverviewFactory(
    * @param {Config} config
    * @return void
    */
-  async function outputStatusOverview(config) {
-    const rows = [];
-
+  async function outputStatusOverview(config, flags) {
```
Comment (shumkov) [inline src/status/outputStatusOverviewFactory.js]: The flags are command logic. I would pass here a specific argument with format to decouple it from commands.

File: src/commands/status/services.js [hunk 5]
```diff
@@ -70,23 +71,37 @@ class ServicesStatusCommand extends ConfigBaseCommand {
...
     }
 
-    const tableConfig = {
-      // singleLine: true,
-      drawHorizontalLine: (index, size) => index === 0 || index === 1 || index === size,
-    };
+    let output;
+
+    if (getFormat(flags) === OUTPUT_FORMATS.JSON) {
+      outputRows.forEach((outputRow, i) => {
```
Comment (shumkov) [inline src/commands/status/services.js]: Let's do array of objects

File: src/commands/status/core.js [hunk 6]
```diff
@@ -172,28 +173,44 @@ class CoreStatusCommand extends ConfigBaseCommand {
...
     }
 
-    const output = table(rows, { singleLine: true });
+    if (insightURLs[config.get('network')]) {
+      outputRows['Remote block height'] = explorerBlockHeight;
+    }
+
+    let output;
+
+    if (getFormat(flags) === OUTPUT_FORMATS.JSON) {
```
Comment (shumkov) [inline src/commands/status/core.js]: I would recommend to extract this logic into function  `printers/printObject(object, format)` to DRY

File: src/commands/status/services.js [hunk 7]
```diff
@@ -70,23 +71,37 @@ class ServicesStatusCommand extends ConfigBaseCommand {
...
       ]);
     }
 
-    const tableConfig = {
-      // singleLine: true,
-      drawHorizontalLine: (index, size) => index === 0 || index === 1 || index === size,
-    };
+    let output;
+
+    if (getFormat(flags) === OUTPUT_FORMATS.JSON) {
```
Comment (shumkov) [inline src/commands/status/services.js]: This one you can extract to `printers/printArrayOfObjects(objects, format)`

**Review by shumkov — CHANGES_REQUESTED** [review]

**Comment by riongull** [comment]
Thanks for this! I will be able to use this with the Incubator Discord bot.

File: src/oclif/command/BaseCommand.js (hunk 3 above)
Comment (strophy) [reply src/oclif/command/BaseCommand.js]: I used `table` instead so we have the option to add real `plain` output later (without table or ansi colours)

Comment (shumkov) [reply src/oclif/command/BaseCommand.js]: I was thinking that plain is just ascii (text).

Comment (strophy) [reply src/oclif/command/BaseCommand.js]: So the format enum contains `[plain, json, table]`?

**Comment by strophy** [comment]
Sample command output: https://gist.github.com/strophy/26fa81c8218bdf76fff7bd49536d062a

File: src/oclif/command/BaseCommand.js (hunk 3 above)
Comment (shumkov) [reply src/oclif/command/BaseCommand.js]: I was thinking that output tables are “plain” (text) format. Not all commands have tables. Maybe I’m wrong..

Comment (strophy) [reply src/oclif/command/BaseCommand.js]: I see. Only status commands have table output, nearly all other commands use listr for output I think? Should we add JSON + plain output options for non-status commands?

**Comment by strophy** [comment]
Closed in favor of https://github.com/dashevo/platform/pull/31/

**Pull request closed without merging**